File: src/modules/stored-file/services/stored-file-update.service.ts
import { prisma } from '../../../infra/database/prisma.service';
import { FileAdapter } from '../../../infra/file-system/adapters/file-adapter';
import { awsFileSystem } from '../../../infra/file-system/implementations/aws.file-system';
import { NotFoundException } from '../../../shared/exceptions';
import { generateFileName } from '../../../shared/utils/file';
import { hashMd5 } from '../../../shared/utils/hash';
import { StoredFileUpdateDto } from '../domain/dto/stored-file-update.dto';
import { StoredFileEntity } from '../domain/entities/stored-file.entity';

class StoredFileUpdateService {
  async execute(
    uuid: string,
    file: FileAdapter,
    dto: StoredFileUpdateDto,
  ): Promise<StoredFileEntity> {
    const storedFile = await prisma.storedFile.findUnique({
      where: { uuid: uuid },
    });

    if (!storedFile) throw new NotFoundException('Arquivo não encontrado!');

    const storedName = generateFileName(dto.name ?? file.name);

    await awsFileSystem.delete(
      `${storedFile.relativePath}/${storedFile.storedName}`,
    );

    await awsFileSystem.put({
      name: storedName,
      relativePath: dto.relativePath ?? storedFile.relativePath,
      content: file.content,
      length: file.length,
      contentType: file.contentType,
    });

    const updatedStoredFile = await prisma.storedFile.update({
      where: { uuid: uuid },
      data: {
        alt: dto.alt,
        checksum: await hashMd5(file.content),
        originalName: file.name,
        storedName: storedName,
        relativePath: dto.relativePath,
        contentType: file.contentType,
        isPublic: dto.isPublic,
      },
    });

    return new StoredFileEntity(updatedStoredFile);
  }
}

export const storedFileUpdateService = new StoredFileUpdateService();

File: src/modules/stored_file/services/stored_file_update_service.py
from src.infra.file_system.implementations.aws_file_system import AwsFileSystem
from src.infra.file_system.adapters.file_adapter import FileAdapter
from src.modules.stored_file.models import StoredFile
from src.modules.stored_file.domain.dto.stored_file_update_dto import (
    StoredFileUpdateDTO,
)
from src.shared.utils.file import generate_file_name
from src.shared.utils.hash import hash_md5
from src.shared.exceptions.not_found_exception import NotFoundException


class StoredFileUpdateService:
    @classmethod
    def execute(self, uuid: str, file: FileAdapter, dto: StoredFileUpdateDTO):
        stored_file = StoredFile.objects.filter(uuid=uuid).first()

        if stored_file is None:
            raise NotFoundException("Arquivo não encontrado!")

        stored_name = generate_file_name(dto["name"] if "name" in dto else file["name"])

        relative_path = (
            dto["relative_path"]
            if "relative_path" in dto
            else stored_file.relative_path
        )

        AwsFileSystem.delete(f"{stored_file.relative_path}/{stored_file.stored_name}")

        AwsFileSystem.put(
            {
                "name": stored_name,
                "relative_path": relative_path,
                "content": file["content"],
                "length": file["length"],
                "content_type": file["content_type"],
            }
        )

        updated_file = StoredFile.objects.filter(uuid=uuid).first()

        updated_file.alt = dto["alt"]
        updated_file.checksum = hash_md5(file["content"])
        updated_file.original_name = file["name"]
        updated_file.stored_name = stored_name
        updated_file.relative_path = relative_path
        updated_file.content_type = file["content_type"]
        updated_file.is_public = (
            dto["is_public"] if "is_public" in dto else stored_file.is_public
        )

        updated_file.save()

        return updated_file
